openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,294 @@
+	removeCPUSetAnnotation(pod.Annotations)
+
+	// update init containers resources
+	updateContainersResources(pod.Spec.InitContainers, pod.Annotations, class)
+
+	// update app containers resources
+	updateContainersResources(pod.Spec.Containers, pod.Annotations, class)
+}
+
+func updateContainersResources(containers []coreapi.Container, podAnnotations map[string]string, podQoSClass coreapi.PodQOSClass) {
+	for i := range containers {
+		c := &containers[i]
+		cpusAnnotationKey := fmt.Sprintf("%s/%s", containerCPUSharesAnnotation, c.Name)
+
+		if podQoSClass == coreapi.PodQOSBestEffort {
+			podAnnotations[cpusAnnotationKey] = "2"
+			continue
+		}
+
+		if c.Resources.Requests != nil {
+			if _, ok := c.Resources.Requests[coreapi.ResourceCPU]; !ok {
+				continue
+			}
+
+			cpuRequest := c.Resources.Requests[coreapi.ResourceCPU]
+			cpuRequestInMilli := cpuRequest.MilliValue()
+
+			cpuShares := cm.MilliCPUToShares(cpuRequestInMilli)
+			podAnnotations[cpusAnnotationKey] = fmt.Sprintf("%d", cpuShares)
+			delete(c.Resources.Requests, coreapi.ResourceCPU)
+
+			if c.Resources.Limits == nil {
+				c.Resources.Limits = coreapi.ResourceList{}
+			}
+
+			// multiple the CPU request by 1000, to make sure that the resource will pass integer validation
+			newCPURequest := resource.NewMilliQuantity(cpuRequestInMilli*1000, cpuRequest.Format)
+			c.Resources.Requests[managedCapacityLabel] = *newCPURequest
+			c.Resources.Limits[managedCapacityLabel] = *newCPURequest
+		}
+	}
+}
+
+func isGuaranteed(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request, requestExist := c.Resources.Requests[r]
+
+			if limit.IsZero() {
+				return false
+			}
+
+			if !requestExist {
+				continue
+			}
+
+			// it some corner case, when you set CPU request to 0 the k8s will change it to the value
+			// specified under the limit
+			if r == coreapi.ResourceCPU && request.IsZero() {
+				continue
+			}
+
+			if !limit.Equal(request) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isBestEffort(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request := c.Resources.Requests[r]
+
+			if !limit.IsZero() || !request.IsZero() {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getPodQoSClass(containers []coreapi.Container) coreapi.PodQOSClass {
+	if isGuaranteed(containers) {
+		return coreapi.PodQOSGuaranteed
+	}
+
+	if isBestEffort(containers) {
+		return coreapi.PodQOSBestEffort
+	}
+
+	return coreapi.PodQOSBurstable
+}
+
+func isManagementResourceAvailable(nodes []*corev1.Node) bool {
```

Comment:
This logic needs to check that *all* nodes are advertising the resource as available. That will ensure that we do not enable the feature until it can be used by all nodes in the cluster. See https://github.com/openshift/enhancements/pull/703/files#diff-bdf426b496fe2c7122db9eeaf303b30ff4561b4f7ac157c21996edff1e67e239R252-R261

Reply:
done